src/main/kotlin/org/dif/peerdid/PeerDIDUtils.kt

```diff
@@ -0,0 +1,149 @@
+        "authentication" to gson.toJsonTree(decodedEncnumbasis)
+    )
+    return gson.toJson(diddoc)
+}
+
+internal fun buildDIDDocNumalgo2(peerDID: PeerDID): String {
+    val keys = peerDID.drop(11)
+    val keysList = keys.split(".")
+    var service = ""
+    val keysWithoutPurposeCode = mutableListOf<String>()
+    for (key in keysList) {
+        if (key[0] != 'S') {
+            val transform = key[1]
+            if (!isInEncodingTypes(transform))
+                throw IllegalArgumentException("Unsupported transform part of PeerDID: $transform")
+
+            keysWithoutPurposeCode.add(key.drop(2))
+        } else service = key.drop(1)
+    }
+    val decodedEncnumbasises = keysWithoutPurposeCode.map { key -> decodeEncnumbasis(key, peerDID) }
+    val decodedService = decodeService(service, peerDID)
+
+    val authentication = mutableListOf<Map<String, String>>()
+    val keyAgreement = mutableListOf<Map<String, String>>()
+
+    for (i in decodedEncnumbasises.indices) {
+        if (PublicKeyTypeAuthentication.values().any { value -> value.name == decodedEncnumbasises[i]["type"] })
+            authentication.add(decodedEncnumbasises[i])
+        else if (PublicKeyTypeAgreement.values().any { value -> value.name == decodedEncnumbasises[i]["type"] })
+            keyAgreement.add(decodedEncnumbasises[i])
+        else
+            throw IllegalArgumentException("Invalid key type of: ${keysList[i]}")
+    }
+
+    val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
+    val diddoc = mapOf(
+        "id" to peerDID,
+        "authentication" to gson.toJsonTree(authentication),
+        "keyAgreement" to gson.toJsonTree(keyAgreement),
+        "service" to gson.toJsonTree(decodedService)
+    )
+    return gson.toJson(diddoc)
+}
+
+private fun isInEncodingTypes(encodingAlgorithm: Char): Boolean {
+    return EncodingType.values().any { type -> type.type == encodingAlgorithm }
+}
+
+private fun decodeService(encodedService: JSON, peerDID: PeerDID): Map<String, String> {
+    val decodedService = Base64.decodeBase64(encodedService).decodeToString()
+    val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
+    val serviceMap = gson.fromJson(decodedService, HashMap::class.java)
+    val serviceType = serviceMap.remove("t").toString().replace("dm", "didcommmessaging")
+
+    return mapOf(
+        "id" to peerDID.plus('#').plus(serviceType),
+        "type" to serviceType,
+        "serviceEndpoint" to serviceMap.remove("s").toString(),
+        "routingKeys" to serviceMap.remove("r").toString()
+    )
+}
+
+private fun decodeEncnumbasis(encnumbasis: String, peerDID: PeerDID): Map<String, String> {
+    val decodedEncnumbasis = Base58.decode(encnumbasis)
+    val codec = getCodec(decodedEncnumbasis)
+    val decodedEncnumbasisWithoutPrefix = removePrefix(decodedEncnumbasis)
+    val publicKey = Base58.encode(decodedEncnumbasisWithoutPrefix.toByteArray())
+    return mapOf(
```

Comment:
I would make it symmetric with `createEncnumbasis` and return just `PublicKey` instance for the following reasons:
1) let's not mix crypto and how it's represented in DIDDoc (exact field names)
2) symmetric with `createEncnumbasis`
3) Makes some other checks easier